### PR TITLE
Feature/fp 65

### PR DESCRIPTION
**Pull request by MichielVrins**

### CODE REVIEW

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        val authenticators = oneginiSDK.oneginiClient.userClient.getAllAuthenticators(userProfile)
+        val authenticator = authenticators.find { it.isPreferred }
+            ?: return callback(Result.failure(SdkError(BIOMETRIC_AUTHENTICATION_NOT_AVAILABLE).pigeonError()))
+
+        if (authenticator.type == OWAuthenticatorType.PIN.toOneginiInt()) {
+            return callback(Result.success(OWAuthenticator(authenticator.id, authenticator.name, authenticator.isRegistered, authenticator.isPreferred, OWAuthenticatorType.PIN)))
+        }
+        if (authenticator.type == OWAuthenticatorType.BIOMETRIC.toOneginiInt()) {
+            return callback(Result.success(OWAuthenticator(authenticator.id, authenticator.name, authenticator.isRegistered, authenticator.isPreferred, OWAuthenticatorType.BIOMETRIC)))
+        }
```
Comment (aleknow) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt:26]: You could use `when` statement and maybe only select `OWAuthenticatorType` which then would be passed in `callback(Result...)`

File: android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt [hunk 2]
```diff
@@ -140,4 +130,18 @@ class AuthenticateUserUseCaseTests {
       SdkErrorAssert.assertEquals(expected, firstValue.exceptionOrNull())
     }
   }
+
+  private fun WhenUserProfileExists() {
```
Comment (aleknow) [inline android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt:134]: function name should start with lower case letter

File: android/src/test/java/com/onegini/mobile/sdk/GetBiometricAuthenticatorUseCaseTests.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        getBiometricAuthenticatorUseCase(profileId, callbackMock)
+
+        val captor = argumentCaptor<Result<OWAuthenticator>>()
+        verify(callbackMock).invoke(captor.capture())
+        assertEquals(captor.firstValue.getOrNull()?.authenticatorType, OWAuthenticatorType.BIOMETRIC)
+    }
+
+
+    private fun WhenUserProfileExists() {
```
Comment (aleknow) [inline android/src/test/java/com/onegini/mobile/sdk/GetBiometricAuthenticatorUseCaseTests.kt:76]: function name should start with lower case letter

File: android/src/test/java/com/onegini/mobile/sdk/GetPreferredAuthenticatorUseCaseTests.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        WhenUserProfileExists()
+
+        getPreferredAuthenticatorUseCase(profileId, callbackMock)
+
+        val captor = argumentCaptor<Result<OWAuthenticator>>()
+        verify(callbackMock).invoke(captor.capture())
+        SdkErrorAssert.assertEquals(GENERIC_ERROR, captor.firstValue.exceptionOrNull())
+    }
+
+    private fun WhenUserProfileExists() {
```
Comment (aleknow) [inline android/src/test/java/com/onegini/mobile/sdk/GetPreferredAuthenticatorUseCaseTests.kt:87]: name should start with lower case letter

**Review by aleknow — CHANGES_REQUESTED** [review]

File: ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift [hunk 5]
```diff
@@ -13,38 +13,29 @@ class AuthenticatorsHandler: BridgeToAuthenticatorsHandlerProtocol {
...
-            completion(.failure(FlutterError(.noUserProfileIsAuthenticated)))
-            return
-        }
 
+    func registerBiometricAuthenticator(_ profile: UserProfile, _ completion: @escaping (Result<Void, FlutterError>) -> Void) {
         // We don't have to check if the authenticator is already registered as the sdk will do that for us.
         let authenticators = SharedUserClient.instance.authenticators(.all, for: profile)
-        guard let authenticator = authenticators.first(where: { $0.identifier == authenticatorId }) else {
-            completion(.failure(FlutterError(.authenticatorNotFound)))
+        guard let authenticator = authenticators.first(where: { $0.type == AuthenticatorType.biometric }) else {
```
Comment (kfa-onewelcome) [inline ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift:20]: ` $0.type == .biometric` should work

Here and all below occurring

File: ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift [hunk 6]
```diff
@@ -57,6 +48,31 @@ class AuthenticatorsHandler: BridgeToAuthenticatorsHandlerProtocol {
...
+        guard let authenticator = SharedUserClient.instance.authenticators(.all, for: userProfile).first(where: {$0.isPreferred}) else {
+            completion(.failure(FlutterError(.authenticatorNotFound)))
+            return
+        }
+        if authenticator.type == AuthenticatorType.biometric {
+            completion(.success(OWAuthenticator(id: authenticator.identifier, name: authenticator.name, isRegistered: authenticator.isRegistered, isPreferred: authenticator.isPreferred, authenticatorType: OWAuthenticatorType.biometric)))
+            return
+        }
+        if authenticator.type == AuthenticatorType.pin {
+            completion(.success(OWAuthenticator(id: authenticator.identifier, name: authenticator.name, isRegistered: authenticator.isRegistered, isPreferred: authenticator.isPreferred, authenticatorType: OWAuthenticatorType.pin)))
```
Comment (kfa-onewelcome) [inline ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift:70]: ` $0.type == .pin`

File: ios/Classes/NativeBridge/Errors/ErrorMapper.swift [hunk 7]
```diff
@@ -95,6 +96,8 @@ enum OneWelcomeWrapperError: Int {
             return "Mobile Authentication is already in progress and can not be performed concurrently."
         case .browserRegistrationNotInProgress:
             return "Browser registration is currently not in progress."
+        case .biometricAuthenticationNotAvailable:
```
Comment (kfa-onewelcome) [inline ios/Classes/NativeBridge/Errors/ErrorMapper.swift:99]: biometric on iOS means: faceid / touchid - are you checking both?

File: ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift [hunk 8]
```diff
@@ -13,38 +13,29 @@ class AuthenticatorsHandler: BridgeToAuthenticatorsHandlerProtocol {
...
-    func deregisterAuthenticator(_ userProfile: UserProfile, _ authenticatorId: String, _ completion: @escaping (Result<Void, FlutterError>) -> Void) {
-        guard let authenticator = SharedUserClient.instance.authenticators(.all, for: userProfile).first(where: {$0.identifier == authenticatorId}) else {
-            completion(.failure(FlutterError(.authenticatorNotFound)))
-            return
-        }
-
-        if authenticator.isRegistered != true {
-            completion(.failure(FlutterError(.authenticatorNotRegistered)))
+    func deregisterBiometricAuthenticator(_ profile: UserProfile, _ completion: @escaping (Result<Void, FlutterError>) -> Void) {
+        guard let authenticator = SharedUserClient.instance.authenticators(.all, for: profile).first(where: {$0.type == AuthenticatorType.biometric}) else {
```
Comment (kfa-onewelcome) [inline ios/Classes/NativeBridge/Handlers/AuthenticatorsHandler.swift:29]: spaces between { ... }

**Review by kfa-onewelcome — APPROVED** [review]

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt [hunk 9]
```diff
@@ -90,25 +87,23 @@ open class PigeonInterface : UserClientApi, ResourceMethodApi {
...
   @Inject
   lateinit var getRedirectUrlUseCase: GetRedirectUrlUseCase
 
-  @Inject
-  lateinit var getRegisteredAuthenticatorsUseCase: GetRegisteredAuthenticatorsUseCase
-
   @Inject
   lateinit var getUserProfilesUseCase: GetUserProfilesUseCase
 
+
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:98]: remove extra newline

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            ?: return callback(Result.failure(SdkError(GENERIC_ERROR).pigeonError()))
+
+        val authenticatorType = when (authenticator.type) {
+            OWAuthenticatorType.PIN.toOneginiInt() -> OWAuthenticatorType.PIN
+            OWAuthenticatorType.BIOMETRIC.toOneginiInt() -> OWAuthenticatorType.BIOMETRIC
+            // This should never happen because we don't support CUSTOM/FIDO authenticators
+            else -> null
+        } ?: return callback(Result.failure(SdkError(GENERIC_ERROR).pigeonError()))
+
+        return  callback(Result.success(OWAuthenticator(authenticator.id, authenticator.name, authenticator.isRegistered, authenticator.isPreferred, authenticatorType)))
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt:30]: why do we not return the Result in this function as with other functions? This should make it more testable?

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt [hunk 11]
```diff
@@ -200,42 +198,47 @@ open class PigeonInterface : UserClientApi, ResourceMethodApi {
...
+    authenticatorType: OWAuthenticatorType,
+    callback: (Result<OWRegistrationResponse>) -> Unit
+  ) {
+    authenticateUserUseCase(profileId, authenticatorType, callback)
   }
 
-  override fun getAuthenticatedUserProfile(callback: (Result<OWUserProfile>) -> Unit) {
-    callback(getAuthenticatedUserProfileUseCase())
+  override fun authenticateUserPreferred(profileId: String, callback: (Result<OWRegistrationResponse>) -> Unit) {
+    authenticateUserUseCase(profileId, null, callback)
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:214]: Why do we not return the Result in this function as was done previously? This should make it more testable?

Comment (Archifer) [reply android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:214]: Don't really like that null is being used as preferred OWAuthenticatorType; can't we extend this and use a specific type to authenticate indicate preferred authentication?

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt [hunk 12]
```diff
@@ -200,42 +198,47 @@ open class PigeonInterface : UserClientApi, ResourceMethodApi {
...
-  override fun changePin(callback: (Result<Unit>) -> Unit) {
-    changePinUseCase(callback)
+  override fun setPreferredAuthenticator(authenticatorType: OWAuthenticatorType, callback: (Result<Unit>) -> Unit) {
+    callback(setPreferredAuthenticatorUseCase(authenticatorType))
   }
 
-  override fun setPreferredAuthenticator(authenticatorId: String, callback: (Result<Unit>) -> Unit) {
-    callback(setPreferredAuthenticatorUseCase(authenticatorId))
+  override fun deregisterBiometricAuthenticator(callback: (Result<Unit>) -> Unit) {
+    deregisterBiometricAuthenticatorUseCase(callback)
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:230]: cant we return Result from this use case?

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt [hunk 13]
```diff
@@ -200,42 +198,47 @@ open class PigeonInterface : UserClientApi, ResourceMethodApi {
...
+    registerBiometricAuthenticatorUseCase(callback)
   }
 
-  override fun registerAuthenticator(authenticatorId: String, callback: (Result<Unit>) -> Unit) {
-    registerAuthenticatorUseCase(authenticatorId, callback)
+  override fun changePin(callback: (Result<Unit>) -> Unit) {
+    changePinUseCase(callback)
   }
 
+
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:241]: unneeded new line

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    }
+}
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/GetPreferredAuthenticatorUseCase.kt:32]: add new line to the end of the file

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/RegisterBiometricAuthenticatorUseCase.kt [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+                        SdkError(
+                            code = oneginiAuthenticatorRegistrationError.errorType,
+                            message = oneginiAuthenticatorRegistrationError.message
+                        ).pigeonError()
+                    )
+                )
+            }
+        })
+    }
+}
```
Comment (Archifer) [inline android/src/main/kotlin/com/onegini/mobile/sdk/flutter/useCases/RegisterBiometricAuthenticatorUseCase.kt:39]: add new line to the end of the file

File: android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt [hunk 16]
```diff
@@ -80,39 +67,42 @@ class AuthenticateUserUseCaseTests {
...
-    whenever(oneginiSdk.oneginiClient.userClient.getRegisteredAuthenticators(eq(UserProfile("QWERTY")))).thenReturn(emptySet())
+  fun `When a valid ProfileId is passed and null for authenticatorType, Then it should call result success with with UserProfile and CustomInfo as json`() {
+    whenUserProfileExists()
+    whenever(
+      oneginiSdk.oneginiClient.userClient.authenticateUser(
+        anyOrNull(),
+        anyOrNull()
+      )
+    ).thenAnswer {
+      it.getArgument<OneginiAuthenticationHandler>(1).onSuccess(UserProfile("QWERTY"), CustomInfo(0, ""))
```
Comment (Archifer) [inline android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt:78]: why not use the private val for profileID you set previously? for QWERTY?

File: android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt [hunk 17]
```diff
@@ -80,39 +67,42 @@ class AuthenticateUserUseCaseTests {
...
+    }
 
-    authenticateUserUseCase("QWERTY", "TEST", callbackMock)
+    authenticateUserUseCase("QWERTY", null, callbackMock)
 
     argumentCaptor<Result<OWRegistrationResponse>>().apply {
       verify(callbackMock).invoke(capture())
-
-      SdkErrorAssert.assertEquals(AUTHENTICATOR_NOT_FOUND, firstValue.exceptionOrNull())
+      val testUser = OWUserProfile("QWERTY")
```
Comment (Archifer) [inline android/src/test/java/com/onegini/mobile/sdk/AuthenticateUserUseCaseTests.kt:85]: why not use the private val for profileID you set previously? for QWERTY?

File: android/src/test/java/com/onegini/mobile/sdk/SetPreferredAuthenticatorUseCaseTests.kt [hunk 18]
```diff
@@ -36,32 +33,50 @@ class SetPreferredAuthenticatorUseCaseTests {
...
+
+    val result = setPreferredAuthenticatorUseCase(OWAuthenticatorType.BIOMETRIC).exceptionOrNull()
+
     SdkErrorAssert.assertEquals(NO_USER_PROFILE_IS_AUTHENTICATED, result)
   }
 
   @Test
-  fun `When an authenticator id is given that is not related to the authenticated user, Then an error should be thrown`() {
+  fun `When an authenticatorType is given that is not related to the authenticated user, Then should reject with AUTHENTICATOR_NOT_FOUND`() {
     whenever(oneginiSdk.oneginiClient.userClient.authenticatedUserProfile).thenReturn(UserProfile("QWERTY"))
```
Comment (Archifer) [inline android/src/test/java/com/onegini/mobile/sdk/SetPreferredAuthenticatorUseCaseTests.kt:46]: why no global private userprofileID var here for "QWERTY"?

**Review by Archifer — CHANGES_REQUESTED** [review]
see comments

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt (hunk 11 above)
Comment (MichielVrins) [reply android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:214]: I initially had the same idea and actually implemented that. The problem with that we actually use this type when getting the authenticators aswel and then preferred shows as a possible type of authenticator and that's not right. 
```
class OWAuthenticator {
  String id;
  String name;
  bool isRegistered;
  bool isPreferred;
  OWAuthenticatorType authenticatorType;
}
```
Does that make sense?

File: android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt (hunk 12 above)
Comment (MichielVrins) [reply android/src/main/kotlin/com/onegini/mobile/sdk/flutter/PigeonInterface.kt:229]: no, we can't beccause it's async.

**Review by Archifer — APPROVED** [review]

**Pull request merged**